csrc/preseg_passes/move_repeat_forward.cpp

```diff
@@ -0,0 +1,480 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_repeat_forward.h>
+
+#include <device_lower/utils.h>
+#include <dispatch.h>
+#include <ir/utils.h>
+#include <logical_domain_map.h>
+#include <ops/all_ops.h>
+#include <scheduler/tools/static_repeat.h>
+
+#include <unordered_map>
+#include <vector>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+std::vector<Val*> getAllPostDominators(TensorView* tv) {
+  std::deque<std::deque<Val*>> all_use_chains =
+      DependencyCheck::getAllUseChains(tv);
+
+  if (all_use_chains.empty()) {
+    return {};
+  }
+
+  if (all_use_chains.size() == 1) {
+    // Skip the first val as it's the given tensor
+    NVF_ERROR_EQ(all_use_chains.at(0).at(0), tv);
+    return {all_use_chains.at(0).begin() + 1, all_use_chains.at(0).end()};
+  }
+
+  std::vector<std::unordered_set<Val*>> all_use_val_sets;
+  all_use_val_sets.reserve(all_use_chains.size());
+
+  for (const auto& use_chain : all_use_chains) {
+    // Skip the first val
+    all_use_val_sets.emplace_back(use_chain.begin() + 1, use_chain.end());
+  }
+
+  const auto& use_chain = all_use_chains.at(0);
```

Comment:
shouldn't this cover all use chaings, instead of only the first one?

Reply:
This is just one of the chains. It doesn't need to be the first one.

At line 49, the intersection of all the chains are computed, which should be post dominators.

Reply:
ah, I mistakenly thought multiple exits would gives us post dominators that's not on one of the use_chain. But that's not a post dominator. :face_with_head_bandage: 

Thanks for bearing with my ignorance.